packages/lodestar/src/sync/backfill/backfill.ts

```diff
@@ -0,0 +1,220 @@
+import {EventEmitter} from "events";
+import PeerId from "peer-id";
+import {StrictEventEmitter} from "strict-event-emitter-types";
+import {computeStartSlotAtEpoch} from "@chainsafe/lodestar-beacon-state-transition";
+import {IBeaconConfig} from "@chainsafe/lodestar-config";
+import {GENESIS_EPOCH} from "@chainsafe/lodestar-params";
+import {phase0, Root, Slot, allForks} from "@chainsafe/lodestar-types";
+import {ErrorAborted, ILogger} from "@chainsafe/lodestar-utils";
+import {List, toHexString} from "@chainsafe/ssz";
+import {IBeaconChain} from "../../chain";
+import {getMinEpochForBlockRequests} from "../../constants";
+import {IBeaconDb} from "../../db";
+import {INetwork, NetworkEvent, PeerAction} from "../../network";
+import {ItTrigger} from "../../util/itTrigger";
+import {PeerSet} from "../../util/peerMap";
+import {shuffleOne} from "../../util/shuffle";
+import {BackfillSyncError, BackfillSyncErrorCode} from "./errors";
+import {verifyBlockProposerSignature, verifyBlockSequence} from "./verify";
+
+/** Default batch size. Same as range sync (2 epochs) */
+const BATCH_SIZE = 64;
+
+export type BackfillSyncModules = {
+  chain: IBeaconChain;
+  db: IBeaconDb;
+  network: INetwork;
+  config: IBeaconConfig;
+  logger: ILogger;
+};
+
+export type BackfillSyncOpts = {
+  batchSize: number;
+};
+
+export enum BackfillSyncEvent {
+  completed = "BackfillSync-completed",
+}
+
+type BackfillSyncEvents = {
+  // slot param is oldest slot synced
+  [BackfillSyncEvent.completed]: (oldestSlotSynced: Slot) => void;
+};
+
+type BackfillSyncEmitter = StrictEventEmitter<EventEmitter, BackfillSyncEvents>;
+
+export class BackfillSync extends (EventEmitter as {new (): BackfillSyncEmitter}) {
+  private readonly chain: IBeaconChain;
+  private readonly network: INetwork;
+  private readonly db: IBeaconDb;
+  private readonly config: IBeaconConfig;
+  private readonly logger: ILogger;
+  private opts: BackfillSyncOpts;
+  /** last trusted block */
+  private anchorBlock: allForks.SignedBeaconBlock | null = null;
```

Comment:
@mpetrunic Removed the `lastFetchedSlot` since you can get that data from the anchorBlock. Let me know if there was a use for `lastFetchedSlot` that I did not notice.

Reply:
I was using that to handle skip slots

Reply:
hhmm, good point. Tho how can we then protect ourselves against nodes that send skip slots when they shouldn't? In range sync there's a two step "verification" process to guard against that. Can we just use a very big batch size and hope it's larger than a range of skipped slots?

Reply:
Well hash chain will fail after which we will go back and try from last anchor block?